language/move-stdlib/nursery/Vault.move

```diff
@@ -0,0 +1,462 @@
+/// process(&read_cap)
+/// ...
+/// fun process(cap: &Vault::ReadCap<Content>) {
+///     let accessor = Vault::read_accessor(cap);
+///     let content = Vault::borrow(accessor);
+///     << do something with `content: &Content` >>
+///     Vault::release_read_accessor(accessor);
+/// }
+/// ```
+///
+/// ## Delegation
+///
+/// Delegation provides the option to delegate the right to acquire a vault capability to other
+/// signers than the owner of the vault. Delegates still need to authenticate themselves using their
+/// signer, similar as the owner does. All security arguments for owners apply also to delegates.
+/// Delegation can be revoked removing previously granted rights from a delegate.
+///
+/// Delegation can be configured to be transitive by granting the right to acquire a delegation capability
+/// to delegates, which can then further delegate access rights.
+///
+/// By default, when a vault is created, it does not support delegation. The owner of the vault
+/// needs to explicitly enable delegation. This allows to create vaults which are not intended for delegation
+/// and one does not need to worry about its misuse.
+///
+/// Example:
+///
+/// ```move
+/// Vault::new(signer, b"My Vault", Content{ ssn: 525659745 });
+/// // Enable delegation for this vault. Only the owning signer can do this.
+/// Vault::enable_delegation<Content>(signer);
+/// ...
+/// // Delegate read capability to some other signer.
+/// let delegate_cap = Vault::acquire_delegate_cap<Content>(signer);
+/// Vault::delegate_read_cap(&delegate_cap, other_signer);
+/// ...
+/// // Other signer can now acquire read cap
+/// let read_cap = Vault::acquire_read_cap<Content>(other_signer);
+/// ...
+/// // The granted capability can be revoked. There is no need to have the other signer for this.
+/// Vault::revoke_read_cap(&delegate_cap, Signer::address_of(other_signer));
+/// ```
+module 0x1::Vault {
+
+    use 0x1::Errors;
+    use 0x1::Event;
+    use 0x1::Option;
+    use 0x1::Signer;
+    use 0x1::Vector;
+
+    // ================================================================================================================
+    // Error reasons
+
+    const EVAULT: u64 = 0;
+    const EDELEGATE: u64 = 1;
+    const EACCESSOR_IN_USE: u64 = 2;
+    const EACCESSOR_INCONSISTENCY: u64 = 3;
+    const EDELEGATE_TO_SELF: u64 = 4;
+    const EDELEGATION_NOT_SUPPORTED: u64 = 5;
+
+
+    // ================================================================================================================
+    // Capabilities
+
+    /// A capability to read the content of the vault. Notice that the capability cannot be
+    /// stored but can be freely copied and dropped.
+    //
+    /// TODO: currently we need to make `Content` droppable even though it is only used as a type index here. This
+    /// should probably be relaxed in the languages, and only types actually used in the data should influence the
+    /// ability requirements.
```

Comment:
I think this is a serious issue right now in the language and prevents idiomatic use. We often (and quite different than in other languages) use type parameters so they just become part of the memory key, but the type itself may never appear in a field, so their abilities are not relevant from a semantic point of view. I think the definition of inference of abilities of a struct type should be based on the abilities of the field types and not of the parameter types. If a field type happens to be a a parameter type, so it be.

In this case, there is no reason why Cap<T> shouldn't be copy+drop if T isn't. 

I understand that one might argue the actual field types are an implementation detail and should not influence the abilities of a struct. This might be a valid argument for other languages in which type parameters do not play a role like they do in Move, where they are used a major design tool for indexing memory.

@sblackshear @tzakian

Reply:
I agree. Phantom types are a fairly common pattern, and several of the nontrivial uses of abilities we have seen so far have run into this problem (e.g, discussion on the type of `emit_event` in the Diem Framework refactoring to use abilities).